test/cli/test_cli.ml

```diff
@@ -8,19 +8,32 @@ let pp_status ppf = function
   | WSTOPPED _ -> Format.fprintf ppf "WSTOPPED _"
 
 let main ocamlformat =
-  let test ?(stdin = "") ~name ~args () =
-    let cmd = Printf.sprintf "%s %s 2>&1" ocamlformat args in
-    let proc = Unix.open_process cmd in
-    let proc_out, proc_in = proc in
+  let ocamlformat =
+    (* Turn into absolute path *)
+    if Filename.is_relative ocamlformat then
+      Filename.concat (Unix.getcwd ()) ocamlformat
+    else
+      ocamlformat
+  in
+  let test ?(env = [||]) ?pwd ?(stdin = "") ~name ~args () =
+    let pp_cd pp = function
+      | Some d -> Format.fprintf pp "cd '%s'; " d
+      | None -> ()
+    in
+    let cmd = Format.asprintf "%a%s %s 2>&1" pp_cd pwd ocamlformat args in
```

Comment:
This adds a lot of complexity to the test harness just to parse `.ocamlformat` files. We could use `spawn` to simplify that, but actually, does this need to be tested at the end to end level?

What do you think about pulling out the function that parses a `.ocamlformat` file and put that in a test harness instead? It looks like it's the main branch of `Conf.read_config_file`. That way, we can just test a string -> config file and remove all of this.

Reply:
Yes. That's something I want to do next and it's required to have ocamlformat as a library.

Here I wanted to test end-to-end because I'm checking the error message given to the user.
I think we should both keep this and have the config parsing unit-tested.

Reply:
It's not too useful to test the same thing at both level. In my experience it's enough to just test the happy path as e2e (and maybe one error to see how it's rendered), but that has not its place in this test harness. We can think how to make an e2e test for file parsing (ocamlformat / ignore / ocp-indent) in a separate PR. (we can probably do it by hand)

Reply:
(you'll need #1148 to link a test suite, I can give you a hand with that)

Reply:
Yes but here I'm not testing that errors are reported or anything. I'm testing the rendering of messages.
I'm testing a single level, the e2e one.
Unit-tests are for an other PR, this one is about error messages.

Reply:
OK to leave the parsing tests elsewhere. But in that case:
- you don't need that many e2e tests for error rendering. one is enough (do not try to have test coverage from e2e tests)
- since the setup of this test is very different from CLI tests (and that it's not a CLI test), it does not need to be executed in this file.
- this test will be easier expressed directly in a dune file, without the intermediate `.ml` / process handling level.

All these steps should make  your tests in this PR a lot simpler.

Reply:
I don't understand what's wrong with the tests.
We need to test messages and their rendering as it can change in the future and we need to know how.

I'm not trying to test the detection of errors, just 4 kinds that are generated from different places.

I think this tests work well with the expect-test output. I tried clean up the code in the last commit, what do you think ? It looks better after splitting.

Reply:
If dune can do more work for us that would be great, I'm not opposed to having duplicates of tests in many places for now. That's a job we need to do later anyway, determine where and how is tested every usecase, and fill the blanks.

Reply:
yes, we can live with extra tests. however I don't think we should try to use this test harness for things it'll do worse than dune. to expand a bit:

- in #1117 we added a way to run ocamlformat with some options and stdin and compare the results.
- in #1126 this was simplified since running a binary with arguments is simple enough that it can be done in-process. expect test are a good enough balance for this.

However, if we need to add features like changing directory, setting environment variables, etc, this is more complicated in-process than using `(setenv)` and `(chdir)` in a dune file.

So I suggest that we:
- leave `test_cli` unchanged in this PR and use it only for things that just pass arguments and don't need to set env/read `.ocamlformat` files
- use manually written dune rules for these new tests, in a style similar to the rules generated in #1117. we don't need a generator for this.

As for `test_cli`, maybe it was a bit overreaching, and the right balance is a set of manually written dune rules (without a generator). There are not that many cases, so we can avoid having this `test_cli` and leverage dune's ability to run processes with arguments and diff files. I'm happy to undo this part after this PR if that works for you.

Reply:
I agree `test_cli` is becoming too complicated (it was from the beginning) and we will have to maintain it.
Instead of going back to vanilla dune (it still lacks features like allowing failing commands) and is a bit harder to setup when we have new needs (especially with no generators), wouldn't it be better to just use better tools ?
We could use `bos` to cleanup process-handling code or use a cram-like tool (eg. `ppx_expect` if it can do that) ?

Reply:
OK, great. We can live with a couple manual rules. I suggest the following plan to move things forward:

- you convert the new tests to plain dune rules (in this PR)
- I undo `test_cli` as plain dune rules (in another PR)
- we try to use craml or mdx later to run the commands (later)

sounds good?

Reply:
I'm not sure converting tests to plain dune rules is good idea because it's pretty hard and time consuming.
This PR won't be merged now because of something else. What about:
- Keep this PR open
- We try to use craml or mdx (in another PR) for the tests currently in master
- Rebase this PR on that
?